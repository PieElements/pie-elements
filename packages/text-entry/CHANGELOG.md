# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

<a name="0.1.0"></a>
# 0.1.0 (2018-01-11)


### Features

* **text-entry:** first commit ([3870745](https://github.com/PieElements/pie-elements/commit/3870745))
