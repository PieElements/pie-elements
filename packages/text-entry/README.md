# text-entry 

## todos..

* [ ] - animate icon
* [ ] - how do langs work in this instance (if lang is es-ES is there a different set of correct responses?)
* [ ] - answerBlankSize 
* [ ] - integers/numbers only 
* [ ] - allow separator/decimal places 
* [ ] - allow negative 
* [ ] - feedback 
* [ ] - color contrast 
* [ ] - outcome calculation 
