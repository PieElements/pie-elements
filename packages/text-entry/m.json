{
  "langs": [
    "en-US",
    "es-ES"
  ],
  "elements": {
    "text-entry": "../.."
  },
  "models": [
    {
      "id": "1",
      "element": "text-entry",
      "defaultLang": "en-US",
      "correctResponses": {
        "award": 100,
        "values": [
          {
            "lang": "en-US",
            "data": [
              "mutt",
              "hound"
            ]
          },
          {
            "lang": "es-ES",
            "data": [
              "chucho",
              "sabueso"
            ]
          }
        ],
        "ignoreWhitespace": true,
        "ignoreCase": true,
        "feedback": {
          "type": "custom",
          "value": [
            {
              "lang": "en-US",
              "data": "correct!"
            },
            {
              "lang": "es-ES",
              "data": "correcto!"
            }
          ]
        },
        "caseSensitive": true
      },
      "incorrectResponses": {
        "award": 0,
        "feedback": {
          "type": "custom",
          "value": "<div>no</div>"
        }
      },
      "model": {
        "answerBlankSize": "5",
        "answerAlignment": "right",
        "allowDecimal": false,
        "allowIntegersOnly": false
      }
    }
  ]
}