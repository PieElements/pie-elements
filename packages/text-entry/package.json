{
  "name": "@pie-elements/text-entry",
  "description": "A pie text-entry component",
  "keywords": [
    "pie",
    "pie-elements"
  ],
  "version": "0.2.4",
  "main": "src/index.js",
  "repository": {
    "type": "git",
    "url": "https://github.com/PieElements/pie-elements.git"
  },
  "publishConfig": {
    "access": "public"
  },
  "license": "ISC",
  "scripts": {
    "client-test": "jest",
    "controller-test": "cd controller && npm test",
    "configure-test": "cd configure && npm test",
    "test": "npm run client-test && npm run controller-test && npm run configure-test"
  },
  "dependencies": {
    "@pie-libs/pie-player-events": "^0.0.1",
    "@pie-libs/render-ui": "^0.1.6",
    "debug": "^3.0.1",
    "lodash": "^4.17.4",
    "material-ui": "^1.0.0-beta.27",
    "material-ui-icons": "^1.0.0-beta.17",
    "react": "^15.6.1",
    "react-addons-css-transition-group": "^15.6.0",
    "react-dom": "^15.6.1",
    "react-number-format": "PieElements/react-number-format#feature/bad-input"
  },
  "devDependencies": {
    "babel": "^6.23.0",
    "babel-cli": "^6.24.1",
    "babel-core": "^6.26.0",
    "babel-jest": "^22.1.0",
    "babel-loader": "^7.1.1",
    "babel-preset-react": "^6.24.1",
    "babel-register": "^6.24.1",
    "jest": "^22.1.1",
    "react-addons-test-utils": "^15.6.0",
    "react-test-renderer": "^15.6.1"
  }
}