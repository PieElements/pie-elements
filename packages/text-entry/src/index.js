import { ModelSetEvent, SessionChangedEvent } from '@pie-libs/pie-player-events'
import ReactTextEntry from './text-entry.jsx';
import debug from 'debug';
import React from 'react';
import ReactDOM from 'react-dom';

const log = debug('pie-elements:text-entry');

export default class TextEntry extends HTMLElement {

  constructor() {
    super();
  }

  set model(m) {
    this._model = m;
    this.render();
    this.dispatchEvent(new ModelSetEvent(
      this.tagName.toLowerCase(),
      false,
      !!this._model
    ));
  }

  set session(s) {
    this._session = s;
    this.render();
  }

  onSessionChanged(value) {
    this._session.value = value;
    log('[onSessionChanged] session: ', this._session);

    this.dispatchEvent(new SessionChangedEvent(
      this.tagName.toLowerCase(),
      false,
    ));
  }

  connectedCallback() {
    this.render();
  }

  render() {
    if (this._model && this._session) {

      const e = React.createElement(ReactTextEntry, {
        model: this._model,
        session: this._session,
        onSessionChanged: this.onSessionChanged.bind(this)
      });

      ReactDOM.render(e, this, () => {
        log('render completed');
      });
    }
  }
} 