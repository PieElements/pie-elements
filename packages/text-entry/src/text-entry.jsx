import React from 'react';
import PropTypes from 'prop-types';
import debug from 'debug';
import classNames from 'classnames';
import { withStyles } from 'material-ui/styles';
import Input from './input';
import { getFormatTag } from './formatting-component';

const log = debug('pie-elements:text-entry');

const TextEntryStyles = {

  'white_on_black': {
    backgroundColor: 'black'
  },
  'black_on_rose': {
    backgroundColor: 'mistyrose'
  }
}

export class TextEntry extends React.Component {

  constructor(props) {
    super(props);
    this.state = {
      value: props.session && props.session.value || ''
    }
  }

  onChange = (event) => {
    clearTimeout(this.state.timeoutId);
    this.setState({ warning: null, timeoutId: null });
    log('[onChange] value: ', event.target.value);
    if (this.state.value !== event.target.value) {
      this.setState({ value: event.target.value }, () => {
        this.props.onValueChanged(this.state.value);
      });
    }
  }

  onBadInput = (data) => {
    const { model } = this.props;
    const warning = model.numbersOnlyWarning || 'Please enter numbers only';
    log('[onBadInput] warning: ', warning);
    const timeoutId = setTimeout(() => {
      this.setState({ warning: null });
    }, 1000);
    this.setState({ warning, timeoutId });
  }

  render() {
    const { session, model, classes } = this.props;
    log('[render] model: ', model);
    const { allowIntegersOnly } = model;
    const { value } = this.state;
    const FormatTag = getFormatTag(model);
    const inputProps = model.allowIntegersOnly ? { onBadInput: this.onBadInput } : {}
    const names = classNames(classes.textEntry, classes[model.colorContrast]);
    return (
      <div className={names}>
        <Input
          dark={model.colorContrast === 'white_on_black'}
          feedback={model.feedback}
          value={value}
          correctness={model.correctness}
          alignment={model.answerAlignment}
          size={model.answerBlankSize}
          onChange={this.onChange}
          inputComponent={FormatTag}
          error={this.state.warning}
          inputProps={inputProps}
          disabled={model.disabled} />
      </div>
    );
  }
}

TextEntry.propTypes = {}

export default withStyles(TextEntryStyles)(TextEntry);