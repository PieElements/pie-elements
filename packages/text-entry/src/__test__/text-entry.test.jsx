import React from 'react';

import renderer from 'react-test-renderer';
import { TextEntry } from '../text-entry.jsx';

jest.mock('@pie-libs/render-ui', () => ({
  indicators: {
    Correct: jest.fn(),
    Incorrect: jest.fn(),
    PartiallyCorrect: jest.fn(),
    NothingSubmitted: jest.fn()
  }
}));

describe('text-entry', () => {

  it('renders', () => {
    const model = {}
    const session = {}
    const classes = {}

    const tree = renderer.create(<TextEntry
      model={model}
      session={session}
      classes={classes}
    />);
    expect(tree).toMatchSnapshot();

  });
})