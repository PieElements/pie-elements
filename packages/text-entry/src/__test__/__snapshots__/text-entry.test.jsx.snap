// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`text-entry renders 1`] = `
<div
  className=""
>
  <div
    className="MuiFormControl-root-27 RawInput-formControl-2"
    onBlur={[Function]}
    onFocus={[Function]}
    required={false}
  >
    <div
      className="RawInput-inputAndIcon-4"
    >
      <div
        className="MuiInput-root-31 RawInput-inputRoot-3 MuiInput-formControl-32 MuiInput-inkbar-33 MuiInput-underline-37"
        onBlur={[Function]}
        onFocus={[Function]}
      >
        <input
          aria-invalid={false}
          aria-required={false}
          autoComplete={undefined}
          autoFocus={undefined}
          className="MuiInput-input-40 MuiInput-inputSingleline-43"
          defaultValue={undefined}
          disabled={false}
          id={undefined}
          name={undefined}
          onChange={[Function]}
          onKeyDown={undefined}
          onKeyUp={undefined}
          placeholder={undefined}
          readOnly={undefined}
          required={undefined}
          rows={undefined}
          type="text"
          value=""
        />
      </div>
    </div>
    <p
      className="MuiFormHelperText-root-46"
    >
      
    </p>
  </div>
</div>
`;
